File: projects/Java/src/benchmarks/BenchmarkSerializationFinal.java
package benchmarks;

import java.util.*;
import org.openjdk.jmh.annotations.*;

import com.chronoxor.proto.*;
import com.chronoxor.proto.fbe.*;

@State(Scope.Benchmark)
public class BenchmarkSerializationFinal
{
    private Account _account;
    private final AccountFinalModel _writer;
    private final AccountFinalModel _reader;

    public BenchmarkSerializationFinal()
    {
        // Create a new account with some orders
        _account = new Account(1, "Test", State.good, new Balance("USD", 1000.0), new Balance("EUR", 100.0), new ArrayList<Order>());
        _account.orders.add(new Order(1, "EURUSD", OrderSide.buy, OrderType.market, 1.23456, 1000.0));
        _account.orders.add(new Order(2, "EURUSD", OrderSide.sell, OrderType.limit, 1.0, 100.0));
        _account.orders.add(new Order(3, "EURUSD", OrderSide.buy, OrderType.stop, 1.5, 10.0));

        // Serialize the account to the FBE stream
        _writer = new AccountFinalModel();
        _writer.serialize(_account);
        assert _writer.verify();

        // Deserialize the account from the FBE stream
        _reader = new AccountFinalModel();
        _reader.attach(_writer.getBuffer());
        assert _reader.verify();
        _reader.deserialize(_account);
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    public void verify()
    {
        // Verify the account
        _writer.verify();
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    public void serialize()
    {
        // Reset FBE stream
        _writer.reset();

        // Serialize the account to the FBE stream
        _writer.serialize(_account);
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    public void deserialize()
    {
        // Deserialize the account from the FBE stream
        _reader.deserialize(_account);
    }
}

File: projects/Kotlin/src/benchmarks/BenchmarkSerializationFinal.kt
package benchmarks

import java.util.*
import org.openjdk.jmh.annotations.*

@State(Scope.Benchmark)
class BenchmarkSerializationFinal
{
    @Suppress("JoinDeclarationAndAssignment")
    private var _account: com.chronoxor.proto.Account
    private var _writer: com.chronoxor.proto.fbe.AccountFinalModel
    private var _reader: com.chronoxor.proto.fbe.AccountFinalModel

    init
    {
        // Create a new account with some orders
        _account = com.chronoxor.proto.Account(1, "Test", com.chronoxor.proto.State.good, com.chronoxor.proto.Balance("USD", 1000.0), com.chronoxor.proto.Balance("EUR", 100.0), ArrayList())
        _account.orders.add(com.chronoxor.proto.Order(1, "EURUSD", com.chronoxor.proto.OrderSide.buy, com.chronoxor.proto.OrderType.market, 1.23456, 1000.0))
        _account.orders.add(com.chronoxor.proto.Order(2, "EURUSD", com.chronoxor.proto.OrderSide.sell, com.chronoxor.proto.OrderType.limit, 1.0, 100.0))
        _account.orders.add(com.chronoxor.proto.Order(3, "EURUSD", com.chronoxor.proto.OrderSide.buy, com.chronoxor.proto.OrderType.stop, 1.5, 10.0))

        // Serialize the account to the FBE stream
        _writer = com.chronoxor.proto.fbe.AccountFinalModel()
        _writer.serialize(_account)
        assert(_writer.verify())

        // Deserialize the account from the FBE stream
        _reader = com.chronoxor.proto.fbe.AccountFinalModel()
        _reader.attach(_writer.buffer)
        assert(_reader.verify())
        _reader.deserialize(_account)
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    fun verify()
    {
        // Verify the account
        _writer.verify()
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    fun serialize()
    {
        // Reset FBE stream
        _writer.reset()

        // Serialize the account to the FBE stream
        _writer.serialize(_account)
    }

    @Benchmark
    @BenchmarkMode(Mode.Throughput)
    fun deserialize()
    {
        // Deserialize the account from the FBE stream
        _reader.deserialize(_account)
    }
}
